pkg/nsenter/nsenter_test.go

```diff
@@ -0,0 +1,264 @@
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Cloneflags: uintptr(cloneFlags),
+	}
+
+	if err := cmd.Start(); err != nil {
+		return -1, err
+	}
+
+	return cmd.Process.Pid, nil
+}
+
+func TestSetNSNilFileHandlerFailure(t *testing.T) {
+	err := setNS(nil, "")
+	assert.NotNil(t, err, "Should fail because file handler is nil")
+}
+
+func TestSetNSUnknownNSTypeFailure(t *testing.T) {
+	file := &os.File{}
+	err := setNS(file, "")
+	assert.NotNil(t, err, "Should fail because unknown ns type")
+}
+
+func TestSetNSWrongFileFailure(t *testing.T) {
+	nsFile, err := ioutil.TempFile("", "wrong-ns-path")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		nsFilePath := nsFile.Name()
+		nsFile.Close()
+		os.Remove(nsFilePath)
+	}()
+
+	err = setNS(nsFile, NSTypeIPC)
+	assert.NotNil(t, err, "Should fail because file is not a namespace")
+}
+
+var testNamespaceList = []Namespace{
+	{
+		Type: NSTypeCGroup,
+	},
+	{
+		Type: NSTypeIPC,
+	},
+	{
+		Type: NSTypeNet,
+	},
+	{
+		Type: NSTypePID,
+	},
+	{
+		Type: NSTypeUTS,
+	},
+}
+
+func testToRunNil() error {
+	return nil
+}
+
+func TestNsEnterEmptyPathAndPIDFromNSListFailure(t *testing.T) {
+	err := NsEnter(testNamespaceList, testToRunNil)
+	assert.NotNil(t, err, "Should fail because neither a path nor a PID"+
+		" has been provided by every namespace of the list")
+}
+
+func TestNsEnterEmptyNamespaceListSuccess(t *testing.T) {
+	err := NsEnter([]Namespace{}, testToRunNil)
+	assert.Nil(t, err, "Should not fail since closure should return nil: %v", err)
+}
+
+func areNSIdentical(nsPath1, nsPath2 string) (bool, error) {
+	evalNSPath1, err := filepath.EvalSymlinks(nsPath1)
+	if err != nil {
+		return false, err
+	}
+
+	evalNSPath2, err := filepath.EvalSymlinks(nsPath2)
+	if err != nil {
+		return false, err
+	}
+
+	return (evalNSPath1 == evalNSPath2), nil
+}
+
+func TestNsEnterSuccessful(t *testing.T) {
+	nsList := testNamespaceList
+	sleepDuration := 60
+
+	cloneFlags := 0
+	for _, ns := range nsList {
+		cloneFlags |= CloneFlagsTable[ns.Type]
+	}
+
+	sleepPID, err := startSleepBinary(sleepDuration, cloneFlags)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if sleepPID > 1 {
+			unix.Kill(sleepPID, syscall.SIGKILL)
+		}
+	}()
+
+	for idx := range nsList {
+		nsList[idx].Path = getNSPathFromPID(sleepPID, nsList[idx].Type)
+		nsList[idx].PID = sleepPID
+	}
+
+	var sleepPIDFromNsEnter int
+
+	testToRun := func() error {
+		sleepPIDFromNsEnter, err = startSleepBinary(sleepDuration, 0)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	err = NsEnter(nsList, testToRun)
+	assert.Nil(t, err, "%v", err)
+
+	defer func() {
+		if sleepPIDFromNsEnter > 1 {
+			unix.Kill(sleepPIDFromNsEnter, syscall.SIGKILL)
+		}
+	}()
+
+	for _, ns := range nsList {
+		nsPathEntered := getNSPathFromPID(sleepPIDFromNsEnter, ns.Type)
+
+		evalExpectedNSPath, err := filepath.EvalSymlinks(ns.Path)
+		if err != nil {
+			evalExpectedNSPath = err.(*os.PathError).Path
```

Comment:
Might be useful to add comments here describing how you are doing the comparison.

Reply:
Yes good point